File: 프로그래머스/Level 2/스킬트리.swift
// https://school.programmers.co.kr/learn/courses/30/lessons/49993
import Foundation

func solution(_ skill:String, _ skill_trees:[String]) -> Int {
    var answer: Int = 0
    
    for tree in skill_trees {
        var que: [Character] = skill.map { $0 }
        var qIdx = 0;
        var flag = true
        
        for t in tree {
            if que.contains(t) {
                if t == que[qIdx] {
                    qIdx += 1
                } else {
                    flag = false
                    break;
                }
            }
        }
        
        if flag {
            answer += 1
        }
    }
    
    return answer
}

File: 프로그래머스/Level 2/스킬트리.cpp
// https://school.programmers.co.kr/learn/courses/30/lessons/49993
#include <string>
#include <vector>
#include <queue>
#include <algorithm>

using namespace std;

int solution(string skill, vector<string> skill_trees) {
    int answer = 0;
    
    for (string cur: skill_trees) {
        queue<char> que;
        int flag = 1;
        
        for (char sk: skill) que.push(sk);
        
        for (char c: cur) {
            if (skill.find(c) != string::npos) {
                if (c == que.front()) {
                    que.pop();
                } else {
                    flag = 0;
                    break;
                }
            } else {
                continue;
            }
        }
        
        if (flag) answer++;
    }
    return answer;
}